### ISSUE TITLE
Nier:Automata: Can I set a button press to execute different actions based on conditionals?

### ISSUE THREAD

**Issue by bryanwillis7** [kozec/sc-controller#318]
I know the title is a bit confusing, let me explain:

Trying to play Nier:Automata via wine on linux. It works, but whoever made the controls for the PC port was a complete moron. Normally dodging attacks on a controller is done by hitting a single button. You hit the button and the game executes a dodge. Simple, right?

On the PC version, to dodge, you have to double tap the direction your moving in. so to dodge forward, doubletap forward, dodge to the left, double tap left, ect... On the steam controller this feels horrible. You have to tap the joystick twice in the direction you're moving which is an akward feeling.

I wanted to work around this and have a single shoulder button press take care of dodging. To do this I'd have to make the dodge button's input change depending on what directional button is being pressed. For example: If I'm running left and I hit the dodge button: it executes the a left movement, just allowing me to dodge left. If I'm running right and I hit the dodge button, it executes a right movement, thus allowing me to dodge right.

TLDR: Can I map a buttonpress to a conditional statement that executes different things depending on on what other buttons are being pressed on the controller?

**Comment by kozec** [comment]
Sounds little complicated, I just leaved right bumper mapped to itself and it dodges fine in Nier :)

Anyway, yes, what you are describing is what modeshift is for. Click on anything and then "Modeshift" in top-left corner and you can change what it does depending which button is pressed. But if you are trying to make it depending on stick position, it may be problematic.

**Comment by bryanwillis7** [comment]
Hm, what do you mean by you "Leave the right bumper mapped to itself"? 

Like what keyboard key did you bind the right bumper to?

And I've bound my stick to keyboard keys. So I should be able to mode shift depending on which arrow key is pressed, shouldn't I?

**Comment by kozec** [comment]
> Hm, what do you mean by you "Leave the right bumper mapped to itself"?

I believe you are mapping controller to keyboard to control Nier, while it supports gamepad input as well. It doesn't even need xinput overrides IIRC. And with gamepad input, dodge control is much less insane.

> So I should be able to mode shift depending on which arrow key is pressed, shouldn't I?

Modeshift buttons are those on Steam Controller, not from emulated keys that you map into. And I don't think I can change that, because using emulated keys as input would probably end in infinite loop.

Anyway, if you can run sc-controller from github, I've just added option to use stick as input for modeshift. But it will require using custom action:

```
mode(
  X < 0,     type("aa"),
  X > 0,     type("dd")
)
```

should create action that, when button is pressed and ...
- ... stick is tilted left, double-presses A key
- ... stick is tilted right, double-presses D key

I believe that should make dodging possible.

**Comment by Vintodrimmer** [comment]
@bryanwillis7
Sorry for interjecting, but what stops you from using the console's keymap? I have completed Nier: Automata half a year ago and dodging was done by a single shoulder button. (I did use custom full press actions though).

On an unrelated note, I highly recommend using mouse look for general gameplay with some combination activating the default joystick view for the `shoot 'em up` sections.

**Comment by bryanwillis7** [comment]
@Vintodrimmer @kozec 

Perhaps there is a misunderstanding of some sort. The reason I'm using this program to begin with is because I'm running Nier:Automata through wine on my linux OS, and the wine version of steam I'm using doesn't seem to recognize the steam controller inputs. I figured the only way I could get this to work properly would be to bind the controller to keyboard inputs. 

On the dropdown menu for this program, should I click the x-box configuration and give it a try? I already tried playing this game with a generic USB controller and none of the inputs got recognized, so I'm slightly skeptical.

**Comment by kozec** [comment]
> I'm running Nier:Automata through wine on my linux OS, and the wine version of steam I'm using doesn't seem to recognize the steam controller inputs

Wine emulates only gamepad, so you may need xinput emulator to get game to recognize inputs. Try using either mine [dumbxinputemu](https://github.com/kozec/dumbxinputemu#dumb-xinput-emulator) or [x360ce](https://boilingsteam.com/a-step-by-step-guide-to-make-controllers-work-in-wine/), AFAIK, Nier works with both.

**Comment by bryanwillis7** [comment]
after a few hours of tinkering I finally got your dumbxinputemu to work(I think).. but.. the joystick and buttons were all weird. The joystick wouldn't even control movement at all. I couldn't figure out how to map anything either.

**Comment by bryanwillis7** [comment]
@kozec 

Going back to your solution from earlier:

Anyway, if you can run sc-controller from github, I've just added option to use stick as input for modeshift. But it will require using custom action:

mode(
  X < 0,     type("aa"),
  X > 0,     type("dd")
)
should create action that, when button is pressed and ...

... stick is tilted left, double-presses A key
... stick is tilted right, double-presses D key
I believe that should make dodging possible.


I get: 'X' is not trigger nor axis

I'm guessing this has something to do with how I've mapped the joystick to act as "Simple DPAD"?

I'm really sorry I keep bothering you. :( this is just unbelievably frustrating..

**Comment by kozec** [comment]
> Pretty much the game's fault. I'm guessing because the game wants me to actually double tap the directional key, and using the macro, it's not registering the brief moment the button gets unpressed.

You can do really weird combinations with macros, but it will probably require some trials and errors.
For example, you can have that bumber to release "A" key, do double tap and then push that "A" again, all with 50ms pauses between using something like:

```
release(KEY_A); sleep(0.05); press(KEY_A); sleep(0.05); release(KEY_A); sleep(0.05); press(KEY_A); sleep(0.05); release(KEY_A); sleep(0.05);  press(KEY_A); 
```

([docs](https://github.com/kozec/sc-controller/wiki/Custom-Action-Examples-and-Explanations#macros))

**Comment by kozec** [comment]
Using stick with modeshift is now released in v0.4.0.2
